File: Lista de exercicio/projetoPYTHON/questao5/questao5.py
class Animal:
    def som(self):
        raise NotImplementedError("Método deve ser implementado na subclasse")

class Cachorro(Animal):
    def som(self):
        return "Au Au"

class Gato(Animal):
    def som(self):
        return "Miau"

class Pato(Animal):
    def som(self):
        return "Quack"

class pinto(Animal):
    def som(self):
        return "piu"

def fazer_som(animais):
    for animal in animais:
        print(f"{animal.__class__.__name__}: {animal.som()}")

def executar_questao5():
    animais = [Cachorro(), Gato(), Pato(), pinto()]
    fazer_som(animais)

File: Lista de exercicio/projetoJAVA/src/projetojava/questao5.java
package projetojava;

abstract class Animal {
    abstract String som();
}

class Cachorro extends Animal {
    @Override
    String som() {
        return "Au Au";
    }
}

class Gato extends Animal {
    @Override
    String som() {
        return "Miau";
    }
}

class Pato extends Animal {
    @Override
    String som() {
        return "Quack";
    }
}

class Pinto extends Animal {
    @Override
    String som() {
        return "Piu";
    }
}

public class questao5 {
    public static void fazerSom(Animal[] animais) {
        for (Animal animal : animais) {
            System.out.println(animal.getClass().getSimpleName() + ": " + animal.som());
        }
    }

    public static void main(String[] args) {
        Animal[] animais = {new Cachorro(), new Gato(), new Pato(), new Pinto()};
        fazerSom(animais);
    }
}
